### PR TITLE
add goal to room

### PR DESCRIPTION
**Pull request by hollesse**

### CODE REVIEW

File: src/main/java/sh/mob/timer/web/Room.java [hunk 1]
```diff
@@ -24,18 +24,41 @@ final class Room {
...
 
-  public void add(Long timer, String user, Instant requested) {
+  public void addTimer(Long timer, String user, Instant requested) {
     var nextUser = findNextUser(user);
     var timerRequest = new TimerRequest(timer, requested, user, nextUser, TimerType.TIMER);
     timerRequests.add(timerRequest);
     sink.tryEmitNext(timerRequest);
   }
 
+  public void setGoal(String text, String user, Instant requested) {
```
Comment (gregorriegler) [inline src/main/java/sh/mob/timer/web/Room.java:43]: I noticed that the argument order does not aligh with the Constructor

File: src/main/java/sh/mob/timer/web/Room.java [hunk 2]
```diff
@@ -24,18 +24,41 @@ final class Room {
   private final List<TimerRequest> timerRequests = new CopyOnWriteArrayList<>();
   private final Sinks.Many<TimerRequest> sink =
```
Comment (gregorriegler) [inline src/main/java/sh/mob/timer/web/Room.java:25]: now that we have two sinks, this name is too generic I think

File: src/main/java/sh/mob/timer/web/RoomApiController.java [hunk 3]
```diff
@@ -98,6 +99,38 @@ public void publishEvent(@PathVariable String roomId, @RequestBody PutTimerReque
...
+              goalRequest.user(),
+              room.name());
+    } else {
+      log.warn("Could not understand PUT goal request for room {}", roomId);
+    }
+  }
+
+  @DeleteMapping("/{roomId:[A-Za-z0-9-_]+}/goal")
+  @ResponseStatus(HttpStatus.ACCEPTED)
+  public void putGoal(@PathVariable String roomId, @RequestBody DeleteGoalRequest deleteGoalRequest) {
```
Comment (gregorriegler) [inline src/main/java/sh/mob/timer/web/RoomApiController.java:122]: sollte deleteGoal heißen

File: src/main/resources/templates/room.html [hunk 4]
```diff
@@ -513,6 +551,25 @@ <h5>Integration with the mob tool</h5>
...
+      console.log('handling event GOAL_REQUEST ' + event.data);
+      let goalRequest = JSON.parse(event.data);
+      let requested = goalRequest["requested"];
+      let newGoal = goalRequest["goal"];
+      let user = goalRequest["user"];
+
+      if (newGoal === null && user === null) {
+        console.log('Resetting state');
+
+        goal = null;
```
Comment (gregorriegler) [inline src/main/resources/templates/room.html:564]: those two state changes are sort of a transaction, no? When we want the one, we want the other too.
Meant are Lines 564 and 565

File: src/main/java/sh/mob/timer/web/Room.java [hunk 5]
```diff
@@ -24,18 +24,41 @@ final class Room {
...
       Sinks.many().replay().latestOrDefault(NULL_TIMER_REQUEST);
+  private Goal currentGoal = Goal.NO_GOAL;
+
+  private final Sinks.Many<Goal> goalRequestSink =
+      Sinks.many().replay().latestOrDefault(Goal.NO_GOAL);
```
Comment (gregorriegler) [inline src/main/java/sh/mob/timer/web/Room.java:30]: I'm new to this reactive style. Do we need both currentGoal and the goalRequestSink to maintain this state?
It feels strange having to keep both in sync.

Comment (hollesse) [reply src/main/java/sh/mob/timer/web/Room.java:31]: Yes we need these. The Goal is the internal State where the sink ist just for the sever sent events.

File: src/main/java/sh/mob/timer/web/Room.java (hunk 1 above)
Comment (hollesse) [reply src/main/java/sh/mob/timer/web/Room.java:44]: fixed

File: src/main/java/sh/mob/timer/web/RoomApiController.java [hunk 6]
```diff
@@ -92,91 +92,63 @@ public void publishEvent(@PathVariable String roomId, @RequestBody PutTimerReque
...
+      log.warn("Could not understand PUT goal request for room {}", roomId);
     }
   }
 
-  private void incrementTimerStatsExceptForTestRoom(Room room, long timer) {
-    if (!Objects.equals(room.name(), SMOKETEST_ROOM_NAME)) {
-      stats.incrementTimer(timer);
+  @DeleteMapping("/{roomId:[A-Za-z0-9-_]+}/goal")
+  @ResponseStatus(HttpStatus.ACCEPTED)
+  public void deleteGoal(@PathVariable String roomId, @RequestBody DeleteGoalRequest deleteGoalRequest) {
```
Comment (simonharrer) [inline src/main/java/sh/mob/timer/web/RoomApiController.java:122]: Why a delete goal? Could be a put to empty goal

File: src/main/resources/templates/room.html [hunk 7]
```diff
@@ -108,13 +110,48 @@
...
+        <div class="btn-group">
+          <button class="btn btn-outline-secondary" type="button" data-bs-toggle="modal" data-bs-target="#goalModal">🎯</button>
+          <button type="button"
+                  class="btn btn-outline-secondary dropdown-toggle dropdown-toggle-split"
+                  data-bs-toggle="dropdown"
+                  aria-expanded="false">
+            <span class="visually-hidden">Toggle Dropdown</span>
+          </button>
+          <ul class="dropdown-menu">
+            <li><a class="dropdown-item" onclick="deleteGoal()">🗑️ Delete current goal</a></li>
```
Comment (simonharrer) [inline src/main/resources/templates/room.html:123]: I would name this "Clear" and basically clear the textbox, which would trigger a change event sending an empty string as a goal. Keep it simple.

**Comment by simonharrer** [comment]
I would not use a model. Instead, I would use https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/contenteditable for the display of the goal, and when it is changed, I would trigger events or update it accordingly. Simple UI.

**Pull request merged**